### ISSUE TITLE
The `--body` or `--body-file` arguments for `erpione dispense` cannot be parsed

### ISSUE THREAD

**Issue by alisianoi** [gematik/erp-e2e-testsuite#4]
#### Version:

```
/ # erpione --version
erpione 0.2.2
```

#### Steps to reproduce:

Step 1: Create a `body.json` file with the following contents:

```
[{"type":"PZN","category":"Arzneimittel","vaccine":false,"standardSize":"NB","supplyForm":"Ampullen","amount":1,"amountUnit":"Stk","pzn":"09385450","name":"Vitadol Gold Beutel 1000 mg","batch":{"lotNumber":"3351037791","expiryDate":"02/05/2024"}}]
```

Step 2: Prescribe and accept a prescription task:

```
/ # erpf replace kvnr --to F133333337 root/examples/gkv_bundle_01.xml
09:05:52.553 INFO  d.g.t.e.c.c.r.KvnrReplacer - Replace KVNR to F133333337 in KbvErpBundle (gkv_bundle_01.xml)

/ # erpione prescribe --env ru ./root/examples/out/F133333337_gkv_bundle_01.xml
2024-04-22 09:07:03,121 [prescription_generator.pyc] INFO : [1|1] create prescription ./root/examples/out/F133333337_gkv_bundle_01.xml
2024-04-22 09:07:14,418 [prescription_generator.pyc] INFO : Created Prescription with Task-ID 160.000.226.268.219.42 and AccessCode 417f0cfbf4bc4d77f6e807d81cd86edd12f092feb476a121433209f0fe87916e
2024-04-22 09:07:14,418 [xlsx_transformer.pyc] INFO : Write summary to /root/examples/out/out/kthon_ru_20240422_090714.xlsx

/ # erpione accept --env ru --taskid 160.000.226.268.219.42
2024-04-22 09:08:05,612 [prescription_handling.pyc] INFO : Accept Prescription 160.000.226.268.219.42 with AccessCode 417f0cfbf4bc4d77f6e807d81cd86edd12f092feb476a121433209f0fe87916e
2024-04-22 09:08:17,600 [prescription_handling.pyc] INFO : Prescription 160.000.226.268.219.42 accepted: {'prescriptionId': '160.000.226.268.219.42', 'prescriptionReference': '160.000.226.268.219.42', 'forKvnr': 'F133333337', 'accessCode': '417f0cfbf4bc4d77f6e807d81cd86edd12f092feb476a121433209f0fe87916e', 'secret': '9196a73780735f89665da5408ac3a3673e6e5947109cbbbb4a9001f0d37eb1e8', 'insurance': {'insuranceType': 'GKV', 'iknr': '109719018', 'name': 'AOK Nordost', 'wop': 'Niedersachsen', 'insurantState': 'Rentner', 'personGroup': 'Nicht gesetzt'}, 'medication': {'type': 'PZN', 'category': 'Arzneimittel', 'vaccine': False, 'standardSize': 'N1', 'supplyForm': 'Pflaster transdermal', 'amount': 19, 'amountUnit': 'Stk', 'pzn': '07765007', 'name': 'NEUPRO 8MG/24H PFT 7 ST'}}
```

Step 3: Try to dispense the eRezept with a different medication:

```
# Either try to use a body file:
/ # erpione dispense --env ru --taskid 160.000.226.268.219.42 --body-file body.json

# Or try to use the command line directly:
/ # erpione dispense --env ru --taskid 160.000.226.268.219.42 --body [{"type":"PZN","category":"Arzneimittel","vaccine":false,"standardSize":"NB","supplyForm":"Ampullen","amount":1,"amountUnit":"Stk","pzn":"093854
50","name":"Vitadol Gold Beutel 1000 mg","batch":{"lotNumber":"3351037791","expiryDate":"02/05/2024"}}]
```

#### Expected outcome

The eRezept with a substituted medication should be dispensed

#### Observed outcome

The medication fails to be parsed:

```
2024-04-22 09:10:16,802 [prescription_handling.pyc] INFO : Dispense Prescription 160.000.226.268.219.42 with Secret 9196a73780735f89665da5408ac3a3673e6e5947109cbbbb4a9001f0d37eb1e8
2024-04-22 09:10:17,245 [rest_client.pyc     ] ERROR: 400: unknown error:
{"type":"INTERNAL","message":"Invalid JSON Body: Cannot deserialize value of type `java.util.ArrayList<de.gematik.test.erezept.primsys.data.PznDispensedMedicationDto>` from String value (token `JsonToken.VALUE_STRING`)\n at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1]"}
2024-04-22 09:10:17,245 [prescription_handling.pyc] ERROR: Prescription 160.000.226.268.219.42 could not be dispensed: 400
```

##### Additional notes

It is possible that my request body from the `body.json` file doesn't closely follow the expected data types:

1. [List of `PznDispensedMedicationDto`](https://github.com/gematik/erp-e2e-testsuite/blob/0bacfa9c933e26a6dc8dd5df9808f6355319f673/primsys-rest/primsys-rest-data/src/main/java/de/gematik/test/erezept/primsys/data/PznDispensedMedicationDto.java#L26)
2. [`PznMedicationBatchDto`](https://github.com/gematik/erp-e2e-testsuite/blob/0bacfa9c933e26a6dc8dd5df9808f6355319f673/primsys-rest/primsys-rest-data/src/main/java/de/gematik/test/erezept/primsys/data/PznMedicationBatchDto.java)
3. [`PznMedicationDto`](https://github.com/gematik/erp-e2e-testsuite/blob/0bacfa9c933e26a6dc8dd5df9808f6355319f673/primsys-rest/primsys-rest-data/src/main/java/de/gematik/test/erezept/primsys/data/PznMedicationDto.java)

However, after a number of attempts and experiments with the contents, I could not come up with a working medication that would be accepted. Could it be that there is a parsing issue or am I missing some mistake in my medication request?

**Comment by alexanderkaiser** [comment]
Hey @alisianoi thank you for reporting. At first glance, I cannot identify the problem.

But I can tell, that your second approach via `--body` is problematic because of proper escaping. That is actually the main reason of having also `--body-file`. 

I have tried to use your body with the scenario player which worked without any issues:

`scenario --env RU resources/gkv_scenario.json`

where the `gkv_scenario.json` is something like

```json
{
  "meta": {
    "description": "Ein einfaches Szenario für Demo-Zwecke",
    "name": "Test Szenario"
  },
  "steps": [
    {
      "type": "create",
      "description": "create an initial prescription",
      "body": {
        "patient": {
          "kvnr": "F133333337"
        },
        "medication": {
          "category": "00",
          "name": "Äppholung"
        }
      }
    },
    {
      "type": "create",
      "description": "create an initial prescription",
      "file": "root/examples/gkv_bundle_01.xml"
    },
    {
      "type": "accept",
      "description": "Akzeptiere Task ${prescriptions[0].taskId}",
      "idx": 0
    },
    {
      "type": "dispense",
      "description": "Dispensiere Task ${accepted[0].taskId} mit Secret ${accepted[0].secret}",
      "body": [
        {
          "type": "PZN",
          "category": "Arzneimittel",
          "vaccine": false,
          "standardSize": "NB",
          "supplyForm": "Ampullen",
          "amount": 1,
          "amountUnit": "Stk",
          "pzn": "09385450",
          "name": "Vitadol Gold Beutel 1000 mg",
          "batch": {
            "lotNumber": "3351037791",
            "expiryDate": "02/05/2024"
          }
        }
      ]
    }
  ]
}
```

containing your body in the "dispense"-step. Could you please give the scenario-player a try

**Comment by alisianoi** [comment]
Hi @alexanderkaiser, thank you for the quick response!

> I have tried to use your body with the scenario player which worked without any issues:

Thanks for the suggested scenario, here is my execution of it:

```
/ # erpione scenario --env ru scenario.json
2024-04-22 11:53:07,296 [scenario_player.pyc ] INFO : Start Scenario 'Test Szenario' on https://erpps-test.dev.gematik.solutions/ in RU
2024-04-22 11:53:07,296 [step.pyc            ] INFO : Execute Step CREATE: create an initial prescription
2024-04-22 11:53:18,143 [context.pyc         ] INFO : Adding data to context.prescriptions: {'prescriptionId': '160.000.226.269.409.61', 'taskId': '160.000.226.269.409.61', 'accessCode': '379074506ff317e35d1375e6995ca53ca3c3e68516ef49dc984aba99337a5820', 'practitioner': {'type': 'Arzt', 'name': 'Dr. Adelheid Ulmenwald', 'id': '6a7f66bc2bb0f4cd76deaa260abbb484', 'officeName': 'Lake Speed', 'hba': '80276001011699901501', 'smcb': '80276001011699901102', 'docNumber': '549303491', 'docNumberType': 'ZANR', 'docQualificationType': 'Arzt', 'bsnr': '380054582', 'phone': '(07478) 6553205', 'email': 'eliana.lindner@xn--klotzei-90a.ch', 'city': 'Neu Tanjahagen', 'postal': '53939', 'street': 'Carl-Duisberg-Platz 3'}, 'patient': {'kvnr': 'F133333337', 'insuranceType': 'GKV', 'firstName': 'Leticia', 'lastName': 'Kesselschläger', 'birthDate': '25.11.1997', 'city': 'Jaydenhagen', 'postal': '90045', 'street': 'Jüch 393'}, 'medication': {'type': 'PZN', 'category': 'Arzneimittel', 'vaccine': False, 'standardSize': 'NB', 'supplyForm': 'Kegel', 'amount': 7, 'amountUnit': 'Stk', 'pzn': '67922832', 'name': 'Äppholung'}, 'medicationRequest': {'dosage': '1-2-1-2-1-0', 'packageQuantity': 2, 'substitutionAllowed': True, 'bvg': True, 'emergencyFee': False}, 'coverage': {'insuranceType': 'GKV', 'iknr': '103724272', 'name': 'BKK Gildemeister SSeidensticker', 'wop': 'Nordbaden', 'insurantState': 'Mitglieder', 'personGroup': 'BVG'}, 'authoredOn': '22/04/2024 11:53:17'}
2024-04-22 11:53:18,145 [step.pyc            ] INFO : Execute Step CREATE: create an initial prescription
2024-04-22 11:53:19,138 [context.pyc         ] INFO : Adding data to context.prescriptions: {'prescriptionId': '160.000.226.269.410.58', 'taskId': '160.000.226.269.410.58', 'accessCode': '3f183a3a2ac3cdf30a5f1716c6ec97372dcdea91b72c141d119ac17c2355a48e', 'practitioner': {'type': 'Arzt', 'name': 'Ben Schulz', 'id': '6a7f66bc2bb0f4cd76deaa260abbb484', 'officeName': 'Hausarztpraxis', 'hba': '80276001011699901501', 'smcb': '80276001011699901102', 'docNumber': '754236701', 'docNumberType': 'LANR', 'docQualificationType': 'Arzt', 'bsnr': '724444400', 'phone': '030321654987', 'email': 'hausarztpraxis@e-mail.de', 'city': 'Berlin', 'postal': '10623', 'street': 'Herbert-Lewin-Platz 2 Erdgeschoss'}, 'patient': {'kvnr': 'K220635158', 'insuranceType': 'GKV', 'firstName': 'Ludger', 'lastName': 'Königsstein', 'birthDate': '22.06.1935', 'city': 'Esens', 'postal': '26427', 'street': 'Blumenweg'}, 'medication': {'type': 'PZN', 'category': 'Arzneimittel', 'vaccine': False, 'standardSize': 'N1', 'supplyForm': 'Pflaster transdermal', 'amount': 7, 'amountUnit': 'Stk', 'pzn': '07765007', 'name': 'NEUPRO 8MG/24H PFT 7 ST'}, 'medicationRequest': {'dosage': '3-2-0-3-0-3-2', 'packageQuantity': 1, 'substitutionAllowed': True, 'bvg': False, 'emergencyFee': False}, 'coverage': {'insuranceType': 'GKV', 'iknr': '109719018', 'name': 'AOK Nordost', 'wop': 'Niedersachsen', 'insurantState': 'Rentner', 'personGroup': 'Nicht gesetzt'}, 'authoredOn': '22/04/2024 11:53:18'}
2024-04-22 11:53:19,140 [step.pyc            ] INFO : Execute Step ACCEPT: Akzeptiere Task 160.000.226.269.409.61
2024-04-22 11:53:19,775 [context.pyc         ] INFO : Adding data to context.accepted: {'prescriptionId': '160.000.226.269.409.61', 'taskId': '160.000.226.269.409.61', 'accessCode': '379074506ff317e35d1375e6995ca53ca3c3e68516ef49dc984aba99337a5820', 'practitioner': {'type': 'Arzt', 'name': 'Dr. Adelheid Ulmenwald', 'id': '6a7f66bc2bb0f4cd76deaa260abbb484', 'officeName': 'Lake Speed', 'hba': '80276001011699901501', 'smcb': '80276001011699901102', 'docNumber': '549303491', 'docNumberType': 'ZANR', 'docQualificationType': 'Arzt', 'bsnr': '380054582', 'phone': '(07478) 6553205', 'email': 'eliana.lindner@xn--klotzei-90a.ch', 'city': 'Neu Tanjahagen', 'postal': '53939', 'street': 'Carl-Duisberg-Platz 3'}, 'patient': {'kvnr': 'F133333337', 'insuranceType': 'GKV', 'firstName': 'Leticia', 'lastName': 'Kesselschläger', 'birthDate': '25.11.1997', 'city': 'Jaydenhagen', 'postal': '90045', 'street': 'Jüch 393'}, 'medication': {'type': 'PZN', 'category': 'Arzneimittel', 'vaccine': False, 'standardSize': 'NB', 'supplyForm': 'Kegel', 'amount': 7, 'amountUnit': 'Stk', 'pzn': '67922832', 'name': 'Äppholung'}, 'medicationRequest': {'dosage': '1-2-1-2-1-0', 'packageQuantity': 2, 'substitutionAllowed': True, 'bvg': True, 'emergencyFee': False}, 'coverage': {'insuranceType': 'GKV', 'iknr': '103724272', 'name': 'BKK Gildemeister SSeidensticker', 'wop': 'Nordbaden', 'insurantState': 'Mitglieder', 'personGroup': 'BVG'}, 'authoredOn': '22/04/2024 11:53:17', 'secret': '1eba9aa2c3e50af69c8f9e42143056b6b4c2d6fe7e74e19ef1d3194494725554'}
2024-04-22 11:53:19,778 [step.pyc            ] INFO : Execute Step DISPENSE: Dispensiere Task 160.000.226.269.409.61 mit Secret 1eba9aa2c3e50af69c8f9e42143056b6b4c2d6fe7e74e19ef1d3194494725554
2024-04-22 11:53:21,837 [context.pyc         ] INFO : Adding data to context.dispensed: {'prescriptionId': '160.000.226.269.409.61', 'taskId': '160.000.226.269.409.61', 'accessCode': '379074506ff317e35d1375e6995ca53ca3c3e68516ef49dc984aba99337a5820', 'practitioner': {'type': 'Arzt', 'name': 'Dr. Adelheid Ulmenwald', 'id': '6a7f66bc2bb0f4cd76deaa260abbb484', 'officeName': 'Lake Speed', 'hba': '80276001011699901501', 'smcb': '80276001011699901102', 'docNumber': '549303491', 'docNumberType': 'ZANR', 'docQualificationType': 'Arzt', 'bsnr': '380054582', 'phone': '(07478) 6553205', 'email': 'eliana.lindner@xn--klotzei-90a.ch', 'city': 'Neu Tanjahagen', 'postal': '53939', 'street': 'Carl-Duisberg-Platz 3'}, 'patient': {'kvnr': 'F133333337', 'insuranceType': 'GKV', 'firstName': 'Leticia', 'lastName': 'Kesselschläger', 'birthDate': '25.11.1997', 'city': 'Jaydenhagen', 'postal': '90045', 'street': 'Jüch 393'}, 'medication': {'type': 'PZN', 'category': 'Arzneimittel', 'vaccine': False, 'standardSize': 'NB', 'supplyForm': 'Kegel', 'amount': 7, 'amountUnit': 'Stk', 'pzn': '67922832', 'name': 'Äppholung'}, 'medicationRequest': {'dosage': '1-2-1-2-1-0', 'packageQuantity': 2, 'substitutionAllowed': True, 'bvg': True, 'emergencyFee': False}, 'coverage': {'insuranceType': 'GKV', 'iknr': '103724272', 'name': 'BKK Gildemeister SSeidensticker', 'wop': 'Nordbaden', 'insurantState': 'Mitglieder', 'personGroup': 'BVG'}, 'authoredOn': '22/04/2024 11:53:17', 'secret': '1eba9aa2c3e50af69c8f9e42143056b6b4c2d6fe7e74e19ef1d3194494725554'}
```

To me, it looks like the "body" from the "dispense" step is completely ignored? I see a different `medication` echoed in the log above. In other words, I think that the body is specified in the scenario but is actually not used when the scenario is played.

> But I can tell, that your second approach via --body is problematic because of proper escaping. That is actually the main reason of having also --body-file.

True. However, could you please confirm that the `--body-file` does _not_ work for you? If it works for you, could you please provide a file that can successfully be parsed?

**Comment by alexanderkaiser** [comment]
Hi @alisianoi 

in fact, the printout in the scenarioplayer seems to be somewhat misleading. Looking into the produced log-file I can see the same error you described.

[22042024-14_28_39.log](https://github.com/gematik/erp-e2e-testsuite/files/15063004/22042024-14_28_39.log)



I'll let you know as soon as I have a solution

**Comment by alexanderkaiser** [comment]
@alisianoi good news, the issue was located in the erpione itself. I've provided the fix via DockerHub to 0.2.2 (without incrementing the version, my bad)

To get the fix you would still need to re-pull the 0.2.2 or latest via `docker pull gematik1/erpione:latest`. You should now also see a more extensive example for this case in `examples/gkv_scenario.json`

The issue was, that I was reading the given body.json as plain string which probably manipulated the json-structure. Now I'm using `json.load(file)` which solves the issue and gives an additional implicit syntax check of body.

Please let me know if the fix works for you

**Comment by alisianoi** [comment]
Thank you @alexanderkaiser for the fix and for the great support with debugging!

The new version of `erpione` worked for me successfully for `--body-file` switch. I could not make the `--body` switch work (even after trying a number of different escapes). This is absolutely good enough for me though! 

For the `scenario`, I see that the replacement medication also gets sent (in the log file) but the replacement medication is not echoed to stdout (only the original medication from the eRezept is echoed). This is also absolutely good enough for me!

Once again, thank you for your work and support. Closing the issue as fixed.
